CLUSTER.md

```diff
@@ -0,0 +1,514 @@
+operations & scaling.
+
+For example:
+- Reliance on primaries (that may fail-over and move around) for quorum makes it impossible to guarantee availability using rack awareness.
+- Node removal has an inherent race condition which can lead to nodes being re-added
+- No single source of truth about who the primary is for a given slot range
+
+Our main goal with this new approach is to provide a more robust solution, with an easy-to-understand
+behavior and predictable failure modes. Beyond this, we also aim to substantially improve the scalability in order to
+be able to easily support a cluster with thousands of shards.
+
+## Design
+
+### Terminology
+
+| Term | Description |
+| --- | --- |
+| Node | A Single redis-server process |
+| Shard | A logical entity comprised of a group of nodes owning the same slots. It is automatically allocated when a new unique slot range is assigned to a node and automatically deallocated when all slots are moved away |
+| Host | A physical / virtual machine hosting the Flotilla processes |
+
+### High Level Overview
+
+Flotilla is composed of two tiers of strongly consistent, consensus-based control plane systems - The Topology Director 
+(TD) and the Failover Coordinator (FC).
+
+The Topology Director is the source of truth for shard existence and metadata, slot range allocation and Failover Coordinator
+assignments. In addition to the above it also propagates failover information to all Failover Coordinators so nodes across 
+the cluster will be able to return an up-to-date CLUSTER SLOTS response.
+
+The Failover Coordinators is the source of truth for Primary/Replica roles. Each Failover Coordinators is assigned a subset 
+of the total nodes in the cluster, receives & propagates heartbeat information from them and enables them to make failover decisions in a
+consistent & timely manner.
+
+Maintaining these two tiers allows scaling the number of nodes while:
+- Maintaining an upper bound over the time required to decide and propagate failover decisions.
+While the time required to propagate the updated primary/replica status depends on the number of Failover Coordinators in the deployment
+the maximum number of these should always be small enough to allow us to guarantee a substantially better SLA than what
+the current implementation provides.
+- Limiting the clustering overhead on individual nodes in the deployment.
+Depending on pub/sub mode, an individual node only needs to maintain a connection to its Failover Coordinator (and primary/replicas).
+
+While Flotilla can be implemented using any sort of strongly consistent system, the reference implementation we propose
+uses Redis to host the Topology Director & Failover Coordinator logic. This allows for different deployment modes depending on 
+the needs and orchestrating abilities of the administrator:
+- Full Blown: Dedicated nodes holding the Topology Directors and multiple Failover Coordinators dividing the data nodes among them.
```

Comment:
for `Full Blown` mode, do we have option that on TD can support multiple clusters?

Reply:
Apologies for the (very) late reply - missed this comment.
It's possible to enhance the spec to support this mode, but not something that's actually planned at this time due to the additional complexity it introduces.
(In any case, it does not allow dividing the compute/storage requirement - it only reduces the control plane overhead when orchestrating lots of small DBs)